### DISCUSSION TITLE
Adding AI to BetterSEQTA+?

### DISCUSSION THREAD

**Discussion by SethBurkart123** [SethBurkart123/EvenBetterSEQTA#74] [Ideas]
I propose adding AI functionality to EvenBetterSEQTA to enhance its capabilities. This will involve integrating machine learning models and natural language processing techniques to provide smarter and more efficient solutions.

**Comment by MEGA-Dawg68** [comment]
What do you mean? like a chat GPT tab or something else making EvenBetterSEQTA AI integrated? This would look epic either way!

**Comment by SethBurkart123** [comment]
For starters, I'm thinking a whole new direct message creator, that looks soo much better (the direct messages would look good for everyone, even in normal seqta). And I could have AI to help with that.

**Comment by MEGA-Dawg68** [comment]
Yes the current one is ok, but could definitely do with a make over. So did you want to have a message creator in direct messages? If so how are we going to get that to work? AI hosting would be really expensive for a free extension

**Comment by SethBurkart123** [comment]
I'm thinking of potentially using one of those free gpt systems? eg. [this](https://github.com/xtekky/gpt4free?tab=readme-ov-file#interference-openai-proxy-api-use-with-openai-python-package)

**Comment by MEGA-Dawg68** [comment]
That is smart. Will look good none the less!

**Comment by SethBurkart123** [comment]
If this feature was to be implemented we would have to find a better way to integrate it that doesn't make teachers freak out. So, I'll close it for now.

**Comment by MEGA-Dawg68** [comment]
What do you mean? I would have though teachers would love a direct message generator, they would get better messages than 'what is 2+2?'  lol

**Comment by SethBurkart123** [comment]
We just have to be careful how it's implemented so that we don't get people like teachers annoyed with it.

**Comment by Jaxx7594** [comment]
@SethBurkart123 You could just add a feature to turn off AI so if teachers get annoyed you can just switch it off. Don't know why they would get annoyed though

**Comment by Jaxx7594** [comment]
Also @SethBurkart123, ive found those sorta free gpt things can be rather unreliable, not sure about this specific one, but its probably unreliable too. I suggest just making the extension freemium and make the AI functionality like, $5 or something. Most people will be able to pay that if they want the feature, and it'll likely cover the cost and even generate some profit.

**Comment by Jaxx7594** [comment]
Ive got some friends who developed their own api for gpt and image generation, i might be able to get you something even cheaper than openai themselves

**Comment by Crazypersonalph** [comment]
What about a GPT window, but the user has to provide their own API key?

**Comment by Crazypersonalph** [comment]
> Also @SethBurkart123, ive found those sorta free gpt things can be rather unreliable, not sure about this specific one, but its probably unreliable too. I suggest just making the extension freemium and make the AI functionality like, $5 or something. Most people will be able to pay that if they want the feature, and it'll likely cover the cost and even generate some profit.

That means we would have to set up accounts, servers, databases, etc; and also worry about securing everything.

We *could*, but *should* we

**Comment by SethBurkart123** [comment]
> > Also @SethBurkart123, ive found those sorta free gpt things can be rather unreliable, not sure about this specific one, but its probably unreliable too. I suggest just making the extension freemium and make the AI functionality like, $5 or something. Most people will be able to pay that if they want the feature, and it'll likely cover the cost and even generate some profit.
> 
> That means we would have to set up accounts, servers, databases, etc; and also worry about securing everything.
> 
> We _could_, but _should_ we

Well I read all of that stuff, there are a few options. First of all, for the main server, we could use https://www.oracle.com/au/cloud/free/ (I've got a computer running on it and it's pretty great)
or maybe even get away with using the free tier of https://supabase.com/ at the start. I have built some apps with supabase in the past, and it is SUPER easy to get users working.

That and maybe doing stuff with https://ai.cloudflare.com/
It uses the open source models which aren't as powerful, but much cheaper and still work great. Another option is using Claude's api.

There's a lot of ways to do it. A freemium is a good idea, the only thing is; would it put some people off the extension? We don't want it being like `oh they just added AI like everyone else`.

**Comment by Crazypersonalph** [comment]
I have my own Enterprise Proxmox Virtual Environment in which I host my VMs on an old laptop, we could use that for the db and server; I also have my own domain name which we could use for API requests, and I have backend experience, networking experience, and cybersecurity experience.

My main problem for adding AI to BetterSEQTA+ is why. A lot of people these days have a habit of shoving AI into places where it doesn't need to be, which just adds extra complexity to the application and doesn't really help anyone.

**Reply by Jaxx7594** [reply]
I think AI built into the direct messages page would be quite useful, ive used ai for writing messages countless times

**Reply by SethBurkart123** [reply]
yeah well that was what I was trying to weight up. Is it worth it? if so where? It'd have to be well planned if it was added.

**Comment by MEGA-Dawg68** [comment]
> I have my own Enterprise Proxmox Virtual Environment in which I host my VMs on an old laptop, we could use that for the db and server; I also have my own domain name which we could use for API requests, and I have backend experience, networking experience, and cybersecurity experience.
> 
> My main problem for adding AI to BetterSEQTA+ is why. A lot of people these days have a habit of shoving AI into places where it doesn't need to be, which just adds extra complexity to the application and doesn't really help anyone.

The way I see it is as long as we don't shove it in peoples faces _cough cough Microsoft_ and make it be able to be turned off and when it is on not in peoples faces, I dont see any issues

**Comment by SethBurkart123** [comment]
Just an interesting thing. For getting free AI that is reliable https://openrouter.ai/ has a whole bunch of completely free AI hosted along with the other paid inference. We could auto sort for free models and then use those to answer the request, meaning the actual AI wouldn't cost us money that we don't have...
<img width="941" alt="image" src="https://github.com/BetterSEQTA/BetterSEQTA-Plus/assets/108050083/5f88ae12-7b69-4ab3-a0b1-a7e9f3a6a0d9">

**Reply by MEGA-Dawg68** [reply]
I have heard of Mistral, but none of the others

**Reply by SethBurkart123** [reply]
You can technically just have some priority lists which choose the best on there. We'd use the Nous: Capybara 7B for example as main, as it's kinda the best out of those for what we'd be generating.

**Reply by MEGA-Dawg68** [reply]
Fair enough, Do we train the AI or do we just set it up? Im down for helping train the AI if you want

**Reply by SethBurkart123** [reply]
We don't have to train it, because the models are good enough already. I'm just trying to think of **where we could integrate it**? If we want to add it anywhere there is text editing boxes, SEQTA uses something called https://ckeditor.com/ckeditor-4/. Except the version that they are on has reached end of life :( The thing is, I can find a way to write text to it with the version 5, but not version 4 which they are currently on...

**Reply by Jaxx7594** [reply]
Why not just replace it with the new ck version? couldnt you just get the new version, hook the old version seqta tries to load, and replace the script? I aint a web dev so i have no idea if this would be hard to do via an extension

**Reply by MEGA-Dawg68** [reply]
Idk if google would even allow 'hijacking' like that.

**Reply by Jaxx7594** [reply]
Well, adblockers are allowed, dont they just modify the ad scripts?

**Reply by SethBurkart123** [reply]
It's a bit more complicated than that sadly. The API's have changed and you'd have to modify minified code which is genuinely very painful. It might be possible, but I wouldn't know where to start with that kind of thing.

**Reply by Jaxx7594** [reply]
Could look at the vencords (discord client mod, has a version for web discord which is an extension) extension code, vencord patches discords code, so that might give some insight on how it'd be possible (https://github.com/Vendicated/Vencord)

**Reply by Jaxx7594** [reply]
[this](https://stackoverflow.com/questions/65147875/chrome-extension-replace-website-javascript-file-via-wildcard) looks like it might work, ima test in a sec

**Reply by Jaxx7594** [reply]
ive looked into it, and it seems there are just so many changes between ck 4 and 5 that you'd have to replace heaps of code to get it working. ima keep tryna get it to work, but i doubt ill be able to

**Reply by Jaxx7594** [reply]
so, ive gotten as far as formatting the file, and finding the parts that arent compatible with ckeditor 5, but since i dont know js i cant figure out how to fix it, and for some reason chatgpt is being an absolute idiot rn and wont listen to what I ask it to do. if any of yall wanna look at it, it starts at line 27691 in [lib.js](https://files.catbox.moe/p7mix2.js). chatgpt works fine to explain what to do to migrate v4 to v5, but is brainless when trying to actually do it. to test the code, you can use [requestly](https://requestly.io/)  (need the app rather than extension to replace with a file) to hook the request. Main issue is that CKEDITOR doesnt exist in v5.

**Reply by Jaxx7594** [reply]
forgot to add this, to replace it with ckeditor 5, use [this](https://files.catbox.moe/prk62p.png) rule

**Reply by Jaxx7594** [reply]
...or the image host aint gonna work... gimme a sec

**Reply by Jaxx7594** [reply]
[here](https://files.catbox.moe/fjle4i.png)

**Reply by Jaxx7594** [reply]
hey uh, if yall need an api key, i can give u 1, i just found one on github lol. obviously not for the extension, but just for free access to gpt-4

**Reply by MEGA-Dawg68** [reply]
> hey uh, if yall need an api key, i can give u 1, i just found one on github lol. obviously not for the extension, but just for free access to gpt-4

Is it someone else's? I think they might be ticked off that some high schoolers are using all their requests lol

**Reply by Jaxx7594** [reply]
They're using it for a therapy website (they really tryna make gpt a therapist lol), so if we wait for them to release for production, they'll prob remove the limit and we'd be able to hide with all the other traffic. plus they're idiots, they put both a gpt token and a firebase token in plaintext and uploaded it to a public github 🤦

**Reply by ar-cyber** [reply]
> We don't have to train it, because the models are good enough already. I'm just trying to think of **where we could integrate it**? If we want to add it anywhere there is text editing boxes, SEQTA uses something called https://ckeditor.com/ckeditor-4/. Except the version that they are on has reached end of life :( The thing is, I can find a way to write text to it with the version 5, but not version 4 which they are currently on...

Might add it to my CME.

**Comment by Zeke0907** [comment]
I'm not great with this stuff but I love Betterseqta+ and all of the others so I would love it if the AI was free but how would it work?

**Reply by SethBurkart123** [reply]
Currently all of the free AI that I could use would either be from trials or from rate-limited API's. If I have some good hardware on hand I might also just run (some) of it myself. Although we can keep in mind stuff exists like huggingchat where we could mix in some requests to them (https://github.com/Soulter/hugging-chat-api). So probably in order to run it, We'd just need to balance out the requests between various free services and selfhosted compute. Otherwise, it'd get pretty expensive.

Now there is always the idea of paid AI etc. But, I'm not so sure people would be able to pay. And even if they could BetterSEQTA+ would get called a "cash grab" (or similar) where people are just like "oh it's to make the creators lots of money, and they don't care about the actual extension that much." I really don't want people coming to that mindset about it.

**Reply by Jaxx7594** [reply]
You could always hint at the user that they can get a free token from chat.openai.com. Cant directly say how though, as i bet that'd be against google tos and def against openai tos

**Reply by SethBurkart123** [reply]
> You could always hint at the user that they can get a free token from chat.openai.com. Cant directly say how though, as i bet that'd be against google tos and def against openai tos

Yeah, but the problem is that sometimes that stuff is blocked as well 🤷.

**Reply by MEGA-Dawg68** [reply]
> > You could always hint at the user that they can get a free token from chat.openai.com. Cant directly say how though, as i bet that'd be against google tos and def against openai tos
> 
> Yeah, but the problem is that sometimes that stuff is blocked as well 🤷.

I didn't know that was against the TOS. I guess I better be watching out for a court summons lol

**Reply by SethBurkart123** [reply]
> > > You could always hint at the user that they can get a free token from chat.openai.com. Cant directly say how though, as i bet that'd be against google tos and def against openai tos
> > 
> > 
> > Yeah, but the problem is that sometimes that stuff is blocked as well 🤷.
> 
> I didn't know that was against the TOS. I guess I better be watching out for a court summons lol

They'll ban you, but that's about it generally.

**Comment by Jones8683** [comment]
Are BetterSEQTA+ and EvenBetterSeqta the same thing? @SethBurkart123 @ar-cyber 
I just noticed that when I open the 'discussions' tab, it says welcome to EvenBetterSeqta Discussions, instead of BetterSEQTA+ like everywhere else 
![image](https://github.com/user-attachments/assets/72340b80-f824-4fbb-9797-2a47f3b6ca27)

**Reply by ar-cyber** [reply]
EvenBetterSEQTA got migrated to BetterSEQTA+.

**Reply by Jones8683** [reply]
Soooo... Even better SEQTA is BetterSEQTA+ now? They are the same? We only have BetterSEQTA+ now? @ar-cyber

**Reply by ar-cyber** [reply]
EvenBetterSEQTA gave way to BetterSEQTA+ because the name "BetterSEQTA+" is more sensible

**Reply by Jones8683** [reply]
Ok

**Reply by Jones8683** [reply]
So we call it Betterseqta+ now

**Reply by Jones8683** [reply]
I understand
